src/Microsoft.TemplateEngine.TemplateLocalizer.Core/TemplateStringExtractor.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using Microsoft.TemplateEngine.TemplateLocalizer.Core.KeyExtractors;
+using Microsoft.TemplateEngine.TemplateLocalizer.Core.TraversalRules;
+
+namespace Microsoft.TemplateEngine.TemplateLocalizer.Core
+{
+    /// <summary>
+    /// Extracts localizable strings from template.json files.
+    /// </summary>
+    internal sealed class TemplateStringExtractor
+    {
+        private readonly JsonDocument _jsonDocument;
+
+        private readonly IJsonKeyCreator _defaultArrayKeyExtractor = new IndexBasedKeyCreator();
+        private readonly IJsonKeyCreator _defaultObjectKeyExtractor = new PropertyNameBasedKeyCreator();
+
+        private readonly List<TemplateString> _extractedStrings = new List<TemplateString>();
+        private readonly HashSet<string> _extractedStringIds = new HashSet<string>();
```

Comment:
I would prefer if this 2 fields were moved into `ExtractStrings` method and passed to `TraverseJsonElements`, I know `TraverseJsonElements` will have ugly a lot of parameters... but it would make this class thread safe, and `.Clear()` in `ExtractStrings` makes me scared :D